core/benches/scheduler.rs

```diff
@@ -0,0 +1,204 @@
+};
+
+#[global_allocator]
+static GLOBAL: Jemalloc = Jemalloc;
+
+// a bench consumer worker that quickly drain work channel, then send a OK back via completed-work
+// channel
+// NOTE: Avoid creating PingPong within bench iter since joining threads at its eol would
+// introducing variance to bench timing.
+#[allow(dead_code)]
+struct PingPong {
+    threads: Vec<std::thread::JoinHandle<()>>,
+}
+
+impl PingPong {
+    fn new<Tx: TransactionWithMeta + Send + Sync + 'static>(
+        work_receivers: Vec<Receiver<ConsumeWork<Tx>>>,
+        completed_work_sender: Sender<FinishedConsumeWork<Tx>>,
+    ) -> Self {
+        let mut threads = Vec::with_capacity(work_receivers.len());
+
+        for receiver in work_receivers {
+            let completed_work_sender_clone = completed_work_sender.clone();
+
+            let handle = std::thread::spawn(move || {
+                Self::service_loop(receiver, completed_work_sender_clone);
+            });
+            threads.push(handle);
+        }
+
+        Self { threads }
+    }
+
+    fn service_loop<Tx: TransactionWithMeta + Send + Sync + 'static>(
+        work_receiver: Receiver<ConsumeWork<Tx>>,
+        completed_work_sender: Sender<FinishedConsumeWork<Tx>>,
+    ) {
+        while let Ok(work) = work_receiver.recv() {
+            if completed_work_sender
+                .send(FinishedConsumeWork {
+                    work,
+                    retryable_indexes: vec![],
+                })
+                .is_err()
+            {
+                // kill this worker if finished_work channel is broken
+                break;
+            }
+        }
+    }
+}
+
+struct BenchEnv<Tx: TransactionWithMeta + Send + Sync + 'static> {
+    #[allow(dead_code)]
+    pingpong_worker: PingPong,
+    filter_1: fn(&[&Tx], &mut [bool]),
+    filter_2: fn(&TransactionState<Tx>) -> PreLockFilterAction,
+    consume_work_senders: Vec<Sender<ConsumeWork<Tx>>>,
+    finished_consume_work_receiver: Receiver<FinishedConsumeWork<Tx>>,
+}
+
+impl<Tx: TransactionWithMeta + Send + Sync + 'static> BenchEnv<Tx> {
+    fn new() -> Self {
+        let num_workers = 4;
+
+        let (consume_work_senders, consume_work_receivers) =
+            (0..num_workers).map(|_| unbounded()).unzip();
+        let (finished_consume_work_sender, finished_consume_work_receiver) = unbounded();
+        let pingpong_worker = PingPong::new(consume_work_receivers, finished_consume_work_sender);
+
+        Self {
+            pingpong_worker,
+            filter_1: Self::test_pre_graph_filter,
+            filter_2: Self::test_pre_lock_filter,
+            consume_work_senders,
+            finished_consume_work_receiver,
+        }
+    }
+
+    fn test_pre_graph_filter(_txs: &[&Tx], results: &mut [bool]) {
+        results.fill(true);
+    }
+
+    fn test_pre_lock_filter(_tx: &TransactionState<Tx>) -> PreLockFilterAction {
+        PreLockFilterAction::AttemptToSchedule
+    }
+}
+
+// TODO - support both schedulers
```

Comment:
Still searching for a satisfying way to reuse `bench_scheduler_impl` code for both schedulers. Compiler won't let me dyn cast concrete schedulers to `Scheduler<>` due to its not object safe. It benches Greedy only atm.

Reply:
macro rules 70951c2572c45bca061fadb4e79209cb30a669ef